File: lib/widgets/tiering/tiering_analytics.dart
import 'package:snabbit_runner/models/tiering/tier_nudge.dart';
import 'package:snabbit_runner/services/mixpanel_setup.dart';
import 'package:snabbit_runner/services/monitoring/monitoring_service_helper.dart';
import 'package:snabbit_runner/utils/enums.dart';
import 'package:snabbit_runner/utils/tracking_events.dart';

/// Analytics for the tiering widgets (`lib/widgets/tiering`).
///
/// Every event fans to **both Mixpanel and CleverTap** via an explicit
/// `targets` list. The tiering event names aren't in the KMP analytics route
/// catalog, so — exactly like the bifrost webview sink — we name both providers
/// here rather than relying on the catalog (whose universal-sink default would
/// otherwise send these to Mixpanel only). See
/// `lib/services/webview/analytics_sink.dart`.
///
/// All firing is guarded: any failure while building the props map or crossing
/// the analytics channel is reported to [MonitoringServiceHelper] and never
/// thrown back at the widget. (The channel itself is already best-effort; this
/// is defence-in-depth around the call site's prop assembly.)
class TieringAnalytics {
  TieringAnalytics._();

  /// Home-nudge render variants (the `render_type` prop on nudge events).
  static const String renderTypeCoinsCard = 'coins_card';
  static const String renderTypeJob = 'job';
  static const String renderTypeThemed = 'themed';

  /// Both providers, mirroring the webview sink's explicit targeting.
  static const List<String> _bothProviders = ['Mixpanel', 'CleverTap'];

  // ── Home nudge (ApplicableTieringNudge + its render variants) ──────────────

  static void nudgeViewed(TierNudge nudge, Tier? tier, String renderType) =>
      _fire(TrackingEvents.tieringNudgeViewed,
          () => _nudgeProps(nudge, tier, renderType));

  static void nudgeClicked(TierNudge nudge, Tier? tier, String renderType) =>
      _fire(TrackingEvents.tieringNudgeClicked,
          () => _nudgeProps(nudge, tier, renderType));

  // ── Snabbit Udaan intro banner ─────────────────────────────────────────────

  static void bannerViewed({required bool showHeaderImage}) =>
      _fire(TrackingEvents.tieringUdaanBannerViewed,
          () => {'show_header_image': showHeaderImage});

  static void bannerClicked({required bool showHeaderImage}) =>
      _fire(TrackingEvents.tieringUdaanBannerClicked,
          () => {'show_header_image': showHeaderImage});

  // ── Drawer "View tier" row ─────────────────────────────────────────────────

  static void drawerViewed(Tier? tier) => _fire(
      TrackingEvents.tieringViewTierDrawerViewed, () => {'tier': _tier(tier)});

  static void drawerClicked(Tier? tier) => _fire(
      TrackingEvents.tieringViewTierDrawerClicked, () => {'tier': _tier(tier)});

  static Map<String, dynamic> _nudgeProps(
    TierNudge nudge,
    Tier? tier,
    String renderType,
  ) =>
      {
        'nudge_name': nudge.nudgeName ?? '',
        'theme': nudge.theme.name,
        'tier': _tier(tier),
        'render_type': renderType,
        'navigation_route': nudge.navigationRoute ?? '',
      };

  static String _tier(Tier? tier) => tier?.name ?? '';

  /// Builds the props and fires the event to both providers, converting any
  /// throw into a [MonitoringServiceHelper] report instead of crashing the UI.
  static void _fire(String event, Map<String, dynamic> Function() buildProps) {
    try {
      MixpanelSetup.logEvent(event, buildProps(), targets: _bothProviders);
    } catch (e, stackTrace) {
      MonitoringServiceHelper.reportError(
        'tiering_analytics_error',
        {'event': event, 'error': e.toString()},
        stackTrace.toString(),
      )
          // Deliberate: reportError is the last-resort sink — if it itself
          // rejects, logging that failure would recurse, so we swallow here only
          // to avoid an unhandled async error (there is nothing left to report to).
          .catchError((_) {});
    }
  }
}

File: shared/src/commonMain/kotlin/com/snabbit/runner/shared/features/tiering/presentation/TieringAnalytics.kt
package com.snabbit.runner.shared.features.tiering.presentation

import com.snabbit.runner.shared.core.analytics.AnalyticsTracker
import com.snabbit.runner.shared.core.analytics.ProviderKeys
import com.snabbit.runner.shared.features.tiering.domain.model.Tier
import com.snabbit.runner.shared.features.tiering.domain.model.TierNudge

/**
 * Tiering analytics — the KMP port of the Flutter `TieringAnalytics`. Every event
 * fans to BOTH Mixpanel and CleverTap via explicit [targets] (the tiering event
 * names aren't in the KMP route catalog, so — like the webview sink — both
 * providers are named here). Event names + prop keys match Dart 1:1.
 */
class TieringAnalytics(private val tracker: AnalyticsTracker) {

    fun nudgeViewed(nudge: TierNudge, tier: Tier?, renderType: String) =
        track(EVENT_NUDGE_VIEWED, nudgeProps(nudge, tier, renderType))

    fun nudgeClicked(nudge: TierNudge, tier: Tier?, renderType: String) =
        track(EVENT_NUDGE_CLICKED, nudgeProps(nudge, tier, renderType))

    fun bannerViewed(showHeaderImage: Boolean) =
        track(EVENT_BANNER_VIEWED, mapOf(PROP_SHOW_HEADER to showHeaderImage))

    fun bannerClicked(showHeaderImage: Boolean) =
        track(EVENT_BANNER_CLICKED, mapOf(PROP_SHOW_HEADER to showHeaderImage))

    // The surface is the profile tier card (not a drawer); the wire event names keep the
    // legacy `..._view_tier_drawer_...` strings for Dart/dashboard parity — see the consts below.
    fun profileTierViewed(tier: Tier?) = track(EVENT_DRAWER_VIEWED, mapOf(PROP_TIER to tierName(tier)))

    fun profileTierClicked(tier: Tier?) = track(EVENT_DRAWER_CLICKED, mapOf(PROP_TIER to tierName(tier)))

    private fun nudgeProps(nudge: TierNudge, tier: Tier?, renderType: String): Map<String, Any?> = mapOf(
        PROP_NUDGE_NAME to nudge.nudgeName.orEmpty(),
        PROP_THEME to nudge.theme.name,
        PROP_TIER to tierName(tier),
        PROP_RENDER_TYPE to renderType,
        PROP_NAV_ROUTE to nudge.navigationRoute.orEmpty(),
    )

    private fun tierName(tier: Tier?): String = tier?.name.orEmpty()

    private fun track(name: String, props: Map<String, Any?>) =
        tracker.track(name, props, targets = BOTH_PROVIDERS)

    companion object {
        const val RENDER_TYPE_COINS_CARD = "coins_card"
        const val RENDER_TYPE_JOB = "job"
        const val RENDER_TYPE_THEMED = "themed"

        private val BOTH_PROVIDERS = setOf(ProviderKeys.MIXPANEL, ProviderKeys.CLEVERTAP)

        private const val EVENT_NUDGE_VIEWED = "tiering_nudge_viewed"
        private const val EVENT_NUDGE_CLICKED = "tiering_nudge_clicked"
        private const val EVENT_BANNER_VIEWED = "tiering_udaan_banner_viewed"
        private const val EVENT_BANNER_CLICKED = "tiering_udaan_banner_clicked"
        private const val EVENT_DRAWER_VIEWED = "tiering_view_tier_drawer_viewed"
        private const val EVENT_DRAWER_CLICKED = "tiering_view_tier_drawer_clicked"

        private const val PROP_NUDGE_NAME = "nudge_name"
        private const val PROP_THEME = "theme"
        private const val PROP_TIER = "tier"
        private const val PROP_RENDER_TYPE = "render_type"
        private const val PROP_NAV_ROUTE = "navigation_route"
        private const val PROP_SHOW_HEADER = "show_header_image"
    }
}
